core/message-handling.go

```diff
@@ -127,12 +162,21 @@ func defaultIncomingMessageHandler(id uint32, log messagelog.MessageLog, config
 
 	applyCommit := makeCommitApplier(collectCommitment)
 	applyPrepare := makePrepareApplier(id, prepareSeq, collectCommitment, handleGeneratedUIMessage, applyCommit)
+	applyReplicaMessage := makeReplicaMessageApplier(applyPrepare, applyCommit)
 	applyRequest := makeRequestApplier(id, n, view, handleGeneratedUIMessage, applyPrepare)
 
+	var processMessage messageProcessor
+	processMessageThunk := func(msg interface{}) (new bool, err error) {
+		// delay evaluation of processMessage variable value
+		return processMessage(msg)
```

Comment:
This "thunk" is needed to construct a `applicableReplicaMessageProcessor` at line 175. If we were simply to pass `processMessage` there, that would be evaluated to `nil`, because it gets assigned later, at line 179. On the other hand, we also need to construct the `applicableReplicaMessageProcessor` before the `messageProcessor`, because the latter eventually depends on it. 

Basically, we have a recursion and a circular dependency here, and this has to be resolved somehow. Since Go uses strict evaluation (applicable order evaluation), we can resort to using this "thunk" and delay the evaluation of `processMessage` variable until it is actually used, after the whole recursive thing is constructed.

I need to think how to properly explain this in a comment...

Reply:
> Basically, we have a recursion and a circular dependency here, and this has to be resolved somehow.

Yes, that's the point. I don't have a simple counter proposal of recursion-free approach, so I'm OK with current code.

Reply:
I'm thinking of adding the following comment to the code: _Due to recursive nature of message processing, an instance of messageProcessor is eventually required for it to be constructed. On the other hand, it is actually invoked only after getting fully constructed. This "thunk" delays evaluation of processMessage variable, thus resolving this circular dependency._

Reply:
... and similar comment above declaration of `applyReplicaMessageThunk` in 8aac667572754dbfc3cfa33abb628be94d0cea37.

Reply:
> 
> 
> I'm thinking of adding the following comment to the code: _Due to recursive nature of message processing, an instance of messageProcessor is eventually required for it to be constructed. On the other hand, it is actually invoked only after getting fully constructed. This "thunk" delays evaluation of processMessage variable, thus resolving this circular dependency._

Thank you. I like this comment.